pkg/service/one2onerestore/worker_validate.go

```diff
@@ -0,0 +1,247 @@
+// Copyright (C) 2025 ScyllaDB
+
+package one2onerestore
+
+import (
+	"cmp"
+	"context"
+	"math/rand/v2"
+	"slices"
+
+	"github.com/pkg/errors"
+	"github.com/scylladb/scylla-manager/v3/pkg/service/backup/backupspec"
+	"github.com/scylladb/scylla-manager/v3/pkg/util/parallel"
+	"golang.org/x/sync/errgroup"
+)
+
+// validateClusters checks if 1-1-restore can be performed on provided clusters.
+func (w *worker) validateClusters(ctx context.Context, locations []LocationInfo, nodeMappings []nodeMapping) error {
+	if err := w.client.VerifyNodesAvailability(ctx); err != nil {
+		return errors.Wrap(err, "all nodes must be available")
+	}
+
+	sourceNodeInfo, targetNodeInfo, err := w.collectNodeInfo(ctx, locations)
+	if err != nil {
+		return errors.Wrap(err, "collect nodes info")
+	}
+
+	sourceNodeInfo, err = applyNodeMapping(sourceNodeInfo, getSourceMappings(nodeMappings))
+	if err != nil {
+		return errors.Wrap(err, "apply node mappings")
+	}
+
+	slices.SortFunc(sourceNodeInfo, compareNodesInfo)
+	slices.SortFunc(targetNodeInfo, compareNodesInfo)
+
+	if err := clustersAreEqual(sourceNodeInfo, targetNodeInfo); err != nil {
+		return errors.Wrap(err, "clusters not equal")
+	}
+
+	return nil
+}
+
+func applyNodeMapping(sourceNodeInfo []nodeInfo, sourceMappings map[node]node) ([]nodeInfo, error) {
+	var result []nodeInfo
+	if len(sourceNodeInfo) != len(sourceMappings) {
+		return nil, errors.Errorf("nodes count (%d) != node count in mappings (%d)", len(sourceNodeInfo), len(sourceMappings))
+	}
+	for _, source := range sourceNodeInfo {
+		target, ok := sourceMappings[node{DC: source.DC, Rack: source.Rack, HostID: source.HostID}]
+		if !ok {
+			return nil, errors.Errorf("mapping for source node (%v) is not found", source)
+		}
+		source.DC, source.Rack, source.HostID = target.DC, target.Rack, target.HostID
+		result = append(result, source)
+	}
+	return result, nil
+}
+
+func compareNodesInfo(a, b nodeInfo) int {
+	return cmp.Or(
+		cmp.Compare(a.DC, b.DC),
+		cmp.Compare(a.Rack, b.Rack),
+		cmp.Compare(a.HostID, b.HostID),
+		cmp.Compare(a.CPUCount, b.CPUCount),
+		cmp.Compare(a.StorageSize, b.StorageSize),
+		slices.Compare(a.Tokens, b.Tokens),
+	)
+}
+
+func clustersAreEqual(sourceNodeInfo, targetNodeInfo []nodeInfo) error {
```

Comment:
Maybe let's rename it from `Equal` to something else, as we don't check the real equality here (e.g. storage size check). It's more about ensuring restore compatibility.